cudamapper/tests/Test_CudamapperIndexBatcher.cu

```diff
@@ -0,0 +1,916 @@
+                                      std::move(device_batch_target_indices)});
+        }
+        expected_batches.push_back({std::move(host_batch),
+                                    std::move(device_batches)});
+    }
+    // skipping query 2, target 0 due to symmetry
+    // skipping query 2, target 1 due to symmetry
+    {
+        // query 2, target 2
+        std::vector<IndexDescriptor> host_batch_query_indices{{19, 1}};
+        std::vector<IndexDescriptor> host_batch_target_indices{{19, 1}};
+        IndexBatch host_batch{std::move(host_batch_query_indices),
+                              std::move(host_batch_target_indices)};
+        std::vector<IndexBatch> device_batches;
+        {
+            // device query 0, device target 0
+            std::vector<IndexDescriptor> device_batch_query_indices{{19, 1}};
+            std::vector<IndexDescriptor> device_batch_target_indices{{19, 1}};
+            device_batches.push_back({std::move(device_batch_query_indices),
+                                      std::move(device_batch_target_indices)});
+        }
+        expected_batches.push_back({std::move(host_batch),
+                                    std::move(device_batches)});
+    }
+
+    test_generate_batches_of_indices(query_indices_per_host_batch,
+                                     query_indices_per_device_batch,
+                                     target_indices_per_host_batch,
+                                     target_indices_per_device_batch,
+                                     query_parser,
+                                     target_parser,
+                                     query_basepairs_per_index,
+                                     target_basepairs_per_index,
+                                     same_query_and_target,
+                                     expected_batches);
+}
+
+TEST(TestCudamapperIndexBatcher, test_generate_batches_of_indices_exceptions)
+{
+    const number_of_indices_t query_indices_per_host_batch                   = 5;
+    const number_of_indices_t query_indices_per_device_batch                 = 2;
+    const number_of_basepairs_t query_basepairs_per_index                    = 10;
+    const std::shared_ptr<const claragenomics::io::FastaParser> query_parser = claragenomics::io::create_kseq_fasta_parser(std::string(CUDAMAPPER_BENCHMARK_DATA_DIR) + "/10_reads.fasta", 1, false);
+
+    number_of_indices_t target_indices_per_host_batch                   = 5;
+    number_of_indices_t target_indices_per_device_batch                 = 2;
+    std::shared_ptr<const claragenomics::io::FastaParser> target_parser = query_parser;
+    number_of_basepairs_t target_basepairs_per_index                    = 10;
+
+    const bool same_query_and_target = true;
+
+    // indices_per_host_batch different
+    target_indices_per_host_batch = 100;
+    ASSERT_THROW(const auto& generated_batches = generate_batches_of_indices(query_indices_per_host_batch,
```

Comment:
This seems rather dangerous - binding a temporary to a `const&`. I would either not assign the return value at all or assign it to `auto generated_batches`.

Reply:
These `ASSERT_THROW` checks are almost redundant at the moment.
If the last check works correctly, all other checks will pass too.
You need to start with identical query and target variables and then reassign only one for check test.

Reply:
I agree about it not being necessary to assign the result at all. Done.

I only partially agree with the second comment. That assumes certain order of checks in the tests function.
Only could edit `generate_batches_of_indices()` so it first checks `indices_per_host_batch`  and then make a bug in check of `indices_per_device_batch` (or delete it completely) and the test would not catch that

Reply:
Ouch... On my second comment: At first I thought we were talking about different things and wanted to explain again what I meant. Now, while trying that, I realized that you did exactly what I wanted to explain as fix. So please disregard my second comment - the way you do it is perfectly fine.
I got confused that all `target_*` variable got initialized separately and did not realize that they actually get initialized with the same value. Sorry.

Optional: Consider to initialize the `target_*` with their `query_*` counterpart, i.e.
```
number_of_indices_t target_indices_per_host_batch = query_indices_per_host_batch;
```
instead of initializing with the same literals.

Reply:
done